File: practice/find_string.py
#!/usr/bin/env python3

def strStr(haystack: str, needle: str):
    idx = 0
    needleIdx = 0

    while (idx < len(haystack)):
        if (haystack[idx] == needle[needleIdx]):
            needleIdx += 1
        else:
            idx -= needleIdx
            needleIdx = 0

        if (needleIdx == len(needle)):
            return idx - (needleIdx - 1)

        idx += 1

    return -1

def main():
    print(strStr('sadbutsad', 'sad'))
    print(strStr('leetcode', 'leeco'))
    print(strStr('mississippi', 'issip'))

if __name__ == '__main__':
    main()

File: practice/find_string.cpp
#include <iostream>
#include <string>

int strStr(std::string haystack, std::string needle) {
    int idx { 0 };
    int needleIdx { 0 };


    while (idx < haystack.size())
    {
        if (haystack[idx] == needle[needleIdx])
        {
            needleIdx += 1;
        } else
        {
            idx -= needleIdx;
            needleIdx = 0;
        }

        if (needleIdx == needle.size())
        {
            return idx - (needleIdx - 1);
        }

        idx += 1;
    }

    return -1;
}

int main()
{
    std::cout << strStr("sadbutsad", "sad") << '\n';
    std::cout << strStr("leetcode", "leeto") << '\n';
    std::cout << strStr("mississippi", "issip") << '\n';

    return 0;
}

